File: azor-the-chatdog-js/src/llm/openrouterClient.ts
/**
 * OpenRouter LLM Client
 *
 * OpenRouter exposes an OpenAI-compatible API, so we reuse the OpenAI SDK
 * and point it at https://openrouter.ai/api/v1.
 */

import OpenAI from 'openai';
import type {
  ILLMClient,
  ILLMChatSession,
  Message,
  LLMResponse,
} from '../types/index.js';
import { validateOpenRouterConfig } from './openrouterValidation.js';

const OPENROUTER_BASE_URL = 'https://openrouter.ai/api/v1';

/**
 * Wrapper for an OpenRouter chat session
 */
class OpenRouterChatSessionWrapper implements ILLMChatSession {
  private client: OpenAI;
  private modelName: string;
  private systemInstruction: string;
  private history: Message[] = [];

  constructor(
    client: OpenAI,
    modelName: string,
    systemInstruction: string,
    initialHistory?: Message[]
  ) {
    this.client = client;
    this.modelName = modelName;
    this.systemInstruction = systemInstruction;
    this.history = initialHistory || [];
  }

  async sendMessage(text: string): Promise<LLMResponse> {
    this.history.push({
      role: 'user',
      parts: [{ text }],
    });

    const messages: OpenAI.Chat.ChatCompletionMessageParam[] = [
      { role: 'system', content: this.systemInstruction },
      ...this.history.map((msg) => ({
        role: msg.role === 'model' ? ('assistant' as const) : ('user' as const),
        content: msg.parts.map((part) => part.text).join(''),
      })),
    ];

    const completion = await this.client.chat.completions.create({
      model: this.modelName,
      messages,
      max_tokens: 4096,
    });

    const responseText = completion.choices[0].message.content?.trim() || '';

    this.history.push({
      role: 'model',
      parts: [{ text: responseText }],
    });

    return { text: responseText };
  }

  getHistory(): Message[] {
    return this.history;
  }
}

/**
 * OpenRouter LLM Client implementation
 */
export class OpenRouterLLMClient implements ILLMClient {
  private client: OpenAI;
  private modelName: string;
  private apiKey: string;

  constructor(modelName: string, apiKey: string) {
    this.modelName = modelName;
    this.apiKey = apiKey;
    this.client = new OpenAI({
      apiKey,
      baseURL: OPENROUTER_BASE_URL,
    });
  }

  static fromEnvironment(): OpenRouterLLMClient {
    const config = validateOpenRouterConfig();
    return new OpenRouterLLMClient(config.modelName, config.openrouterApiKey);
  }

  createChatSession(
    systemInstruction: string,
    history?: Message[],
    _thinkingBudget?: number
  ): ILLMChatSession {
    return new OpenRouterChatSessionWrapper(
      this.client,
      this.modelName,
      systemInstruction,
      history
    );
  }

  countHistoryTokens(history: Message[]): number {
    let totalTokens = 0;
    for (const msg of history) {
      for (const part of msg.parts) {
        totalTokens += Math.ceil(part.text.length / 4);
      }
    }
    return totalTokens;
  }

  getModelName(): string {
    return this.modelName;
  }

  isAvailable(): boolean {
    return !!this.apiKey && this.apiKey.length > 0;
  }

  preparingForUseMessage(): string {
    return `Preparing OpenRouter model ${this.modelName}...`;
  }

  readyForUseMessage(): string {
    const maskedKey = this.apiKey
      ? `${this.apiKey.substring(0, 8)}...${this.apiKey.substring(this.apiKey.length - 4)}`
      : 'NOT SET';
    return `OpenRouter ${this.modelName} ready (API Key: ${maskedKey})`;
  }
}

File: azor-the-chatdog-py/src/llm/openrouter_client.py
"""
OpenRouter LLM Client Implementation

OpenRouter exposes an OpenAI-compatible API, so we reuse the openai SDK
and point it at https://openrouter.ai/api/v1.
"""

import os
import sys
from typing import Optional, List, Any, Dict
from openai import OpenAI
from dotenv import load_dotenv
from cli import console
from .openrouter_validation import OpenRouterConfig

OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1"


class OpenRouterResponse:
    """Response object with a .text attribute containing the response text."""

    def __init__(self, text: str):
        self.text = text


class OpenRouterChatSession:
    """
    Chat session wrapper for OpenRouter.
    Manages conversation history and maps the universal format to the OpenAI-compatible API.
    """

    def __init__(
        self,
        openrouter_client: OpenAI,
        model_name: str,
        system_instruction: str,
        history: Optional[List[Dict]] = None,
    ):
        self.openrouter_client = openrouter_client
        self.model_name = model_name
        self.system_instruction = system_instruction
        self._history = history or []

    def send_message(self, text: str) -> Any:
        """
        Sends a message to the OpenRouter model and returns a response object.

        Args:
            text: User's message

        Returns:
            OpenRouterResponse with .text attribute
        """
        self._history.append({"role": "user", "parts": [{"text": text}]})

        openai_messages = self._convert_to_openai_format()

        try:
            completion = self.openrouter_client.chat.completions.create(
                model=self.model_name,
                messages=openai_messages,
                max_tokens=4096,
            )

            response_text = completion.choices[0].message.content.strip()

            self._history.append({"role": "model", "parts": [{"text": response_text}]})

            return OpenRouterResponse(response_text)

        except Exception as e:
            console.print_error(f"Error while generating OpenRouter response: {e}")
            error_text = "Sorry, an error occurred while generating a response."
            self._history.append({"role": "model", "parts": [{"text": error_text}]})
            return OpenRouterResponse(error_text)

    def get_history(self) -> List[Dict]:
        """Returns the current conversation history in universal format."""
        return self._history

    def _convert_to_openai_format(self) -> List[Dict[str, str]]:
        """Converts universal history format to OpenAI message format."""
        openai_messages = []

        if self.system_instruction:
            openai_messages.append({"role": "system", "content": self.system_instruction})

        for message in self._history:
            role = message["role"]
            text = message["parts"][0]["text"] if message["parts"] else ""
            openai_role = "assistant" if role == "model" else "user"
            openai_messages.append({"role": openai_role, "content": text})

        return openai_messages


class OpenRouterClient:
    """
    OpenRouter LLM Client.
    Provides the same interface as other LLM clients in this codebase.
    """

    def __init__(self, model_name: str, api_key: str):
        if not api_key:
            raise ValueError("API key cannot be empty or None")

        self.model_name = model_name
        self.api_key = api_key
        self._client = self._initialize_client()

    @staticmethod
    def preparing_for_use_message() -> str:
        return "🤖 Preparing OpenRouter client..."

    @classmethod
    def from_environment(cls) -> 'OpenRouterClient':
        load_dotenv()

        config = OpenRouterConfig(
            model_name=os.getenv('OPENROUTER_MODEL_NAME') or os.getenv('MODEL_NAME', 'qwen/qwen3.5-9b'),
            openrouter_api_key=os.getenv('OPENROUTER_API_KEY', ''),
        )

        return cls(model_name=config.model_name, api_key=config.openrouter_api_key)

    def _initialize_client(self) -> OpenAI:
        try:
            return OpenAI(api_key=self.api_key, base_url=OPENROUTER_BASE_URL)
        except Exception as e:
            console.print_error(f"Error initializing OpenRouter client: {e}")
            sys.exit(1)

    def create_chat_session(
        self,
        system_instruction: str,
        history: Optional[List[Dict]] = None,
        thinking_budget: int = 0,
    ) -> OpenRouterChatSession:
        if not self._client:
            raise RuntimeError("LLM client not initialized")

        return OpenRouterChatSession(
            openrouter_client=self._client,
            model_name=self.model_name,
            system_instruction=system_instruction,
            history=history or [],
        )

    def count_history_tokens(self, history: List[Dict]) -> int:
        if not history:
            return 0

        try:
            text_parts = []
            for message in history:
                if "parts" in message and message["parts"]:
                    text_parts.append(message["parts"][0]["text"])
            return len(" ".join(text_parts)) // 4
        except Exception as e:
            console.print_error(f"Error while counting tokens: {e}")
            return 0

    def get_model_name(self) -> str:
        return self.model_name

    def is_available(self) -> bool:
        return self._client is not None and bool(self.api_key)

    def ready_for_use_message(self) -> str:
        if len(self.api_key) <= 8:
            masked_key = "****"
        else:
            masked_key = f"{self.api_key[:8]}...{self.api_key[-4:]}"

        return f"✅ OpenRouter client ready (Model: {self.model_name}, Key: {masked_key})"

    @property
    def client(self):
        return self._client
